// @ionic/cli-utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROJECT_FILE = 'ionic.config.json';
exports.PROJECT_FILE_LEGACY = 'ionic.project';
exports.PROJECT_TYPES = ['angular', 'ionic-angular', 'ionic1', 'custom'];
var CommandGroup;
(function (CommandGroup) {
    CommandGroup[CommandGroup["Deprecated"] = 0] = "Deprecated";
    CommandGroup[CommandGroup["Hidden"] = 1] = "Hidden";
    CommandGroup[CommandGroup["Beta"] = 2] = "Beta";
})(CommandGroup = exports.CommandGroup || (exports.CommandGroup = {}));
var NamespaceGroup;
(function (NamespaceGroup) {
    NamespaceGroup[NamespaceGroup["Deprecated"] = 0] = "Deprecated";
    NamespaceGroup[NamespaceGroup["Hidden"] = 1] = "Hidden";
    NamespaceGroup[NamespaceGroup["Beta"] = 2] = "Beta";
})(NamespaceGroup = exports.NamespaceGroup || (exports.NamespaceGroup = {}));
var OptionGroup;
(function (OptionGroup) {
    OptionGroup[OptionGroup["Advanced"] = 0] = "Advanced";
    OptionGroup[OptionGroup["AppScripts"] = 1] = "AppScripts";
    OptionGroup[OptionGroup["Cordova"] = 2] = "Cordova";
    OptionGroup[OptionGroup["Deprecated"] = 3] = "Deprecated";
    OptionGroup[OptionGroup["Hidden"] = 4] = "Hidden";
})(OptionGroup = exports.OptionGroup || (exports.OptionGroup = {}));
